freedomofpress/securedrop
# API changes for listing/marking source conversation items that have been seen by journalists

## Status

Ready for review

## Description of Changes
    
Add new API endpoint for listing or marking source conversation items that have been seen by a journalist.
    
Add utility method to mark a heterogeneous collection of Submission and Reply objects seen.
    
Add Submission.is_file and Submission.is_message to encapsulate the characterization based on filename.

Fixes #5475.

## Testing

- Run the dev server with `make dev`.

- Get an API token as the `journalist` user:
  ```
  export TOKEN=$(curl -o- -X POST -H "Content-Type: application/json" --data '{"username":"journalist","passphrase":"correct horse battery staple profanity oil chewy","one_time_code":"047889"}' http://localhost:8081/api/v1/token | jq -r .token)
  ```

- List current submissions: 
  ```
  curl -X GET -H "Content-Type: application/json" -H "Authorization: Token $TOKEN" http://localhost:8081/api/v1/submissions
  ```
  All should have a `seen_by` property, currently an empty list.

- Mark a message seen:
  ```
  curl -X POST -H "Content-Type: application/json" -H "Authorization: Token $TOKEN" --data '{"messages": ["e711d29d-d0a1-440f-80e4-6642b77ec3ea"]}' http://localhost:8081/api/v1/seen
  ```
  Replace the UUID with one from one of the messages in the response from the `/submissions` endpoint, of course.
- Mark the same message seen again. There should be no error.
- Retrieve `/submissions` again. The message's `seen_by` list should now contain your journalist account's UUID.
- Visit the source of that message in the journalist interface. The message you marked should not be listed in bold text. Clicking "Select unread" should not select that message.
- Visit the source list in the journalist interface. Select the source of the message you marked and click "Download Unread". The zip file you receive should not contain the seen message.

- Via the source interface, upload a file, then retrieve `/submissions` again, and mark the file seen:
  ```
  curl -X POST -H "Content-Type: application/json" -H "Authorization: Token $TOKEN" --data '{"files": ["4fe4b7f5-3031-4651-9092-888951681cf7"]}' http://localhost:8081/api/v1/seen
  ```
  Replace the UUID with that of the file you just uploaded, of course.
- Repeat the checks in the journalist interface that you performed for a seen message, confirming that the file you marked seen is always considered read.

- Get another API token, this time as the `dellsberg` journalist account.
- Retrieve the list of all replies:
  ```
  curl -X GET -H "Content-Type: application/json" -H "Authorization: Token $TOKEN" http://localhost:8081/api/v1/replies
  ```
- Mark one of them read: 
  ```
  curl -X POST -H "Content-Type: application/json" -H "Authorization: Token $TOKEN" --data '{"replies": ["8924a691-3b0f-45c1-bfed-b0b91e264855"]}' http://localhost:8081/api/v1/seen
  ```
- Hit `/replies` again and confirm that the `dellsberg` account UUID appears in the reply's `seen_by` list.

## Deployment

Depends on the database changes in #5505.

## Checklist

### If you made changes to the server application code:


### If you made non-trivial code changes:
